File: backend/old/src/utils/pathValidator.ts
import path from 'path';
import fs from 'fs-extra';
import { logger } from '../core/logger';
import { PathValidationResult } from '../types/pathCache';

/**
 * 路径验证工具类
 * 提供路径有效性检查和规范化功能
 */
export class PathValidator {
  /**
   * 验证路径是否有效
   * @param inputPath 要验证的路径
   * @param shouldExist 路径是否应该存在
   * @returns 验证结果
   */
  static async validatePath(inputPath: string, shouldExist: boolean = true): Promise<PathValidationResult> {
    try {
      // 检查路径是否为空
      if (!inputPath || inputPath.trim() === '') {
        return {
          isValid: false,
          error: '路径不能为空',
          exists: false,
          isDirectory: false
        };
      }

      // 获取绝对路径
      const absolutePath = path.resolve(inputPath.trim());

      // 检查路径格式是否合法
      if (!this.isValidPathFormat(absolutePath)) {
        return {
          isValid: false,
          error: '路径格式不合法',
          exists: false,
          isDirectory: false,
          absolutePath
        };
      }

      // 检查路径是否存在
      const exists = await fs.pathExists(absolutePath);

      if (!exists) {
        return {
          isValid: !shouldExist,
          error: shouldExist ? '路径不存在' : undefined,
          exists: false,
          isDirectory: false,
          absolutePath
        };
      }

      // 检查是否为目录
      const stats = await fs.stat(absolutePath);
      const isDirectory = stats.isDirectory();

      if (!isDirectory) {
        return {
          isValid: false,
          error: '路径不是一个目录',
          exists: true,
          isDirectory: false,
          absolutePath
        };
      }

      return {
        isValid: true,
        exists: true,
        isDirectory: true,
        absolutePath
      };

    } catch (error) {
      logger.error('路径验证失败:', error);
      return {
        isValid: false,
        error: `路径验证失败: ${error instanceof Error ? error.message : String(error)}`,
        exists: false,
        isDirectory: false
      };
    }
  }

  /**
   * 检查路径格式是否合法
   * @param pathStr 路径字符串
   * @returns 是否合法
   */
  private static isValidPathFormat(pathStr: string): boolean {
    try {
      // Windows路径特殊字符检查
      if (process.platform === 'win32') {
        // 检查路径中不允许的字符（排除冒号，因为驱动器号需要冒号）
        const invalidChars = /[<>"|?*]/;
        if (invalidChars.test(pathStr)) {
          return false;
        }
        
        // 检查是否包含驱动器号后的无效冒号
        const drivePattern = /^[A-Za-z]:/;
        if (!drivePattern.test(pathStr)) {
          // 如果不是以驱动器号开头，则不能包含冒号
          if (pathStr.includes(':')) {
            return false;
          }
        } else {
          // 如果以驱动器号开头，检查后续部分是否包含冒号
          const afterDrive = pathStr.substring(2);
          if (afterDrive.includes(':')) {
            return false;
          }
        }
      }

      // 通用路径长度检查
      if (pathStr.length > 260 && process.platform === 'win32') {
        return false; // Windows路径长度限制
      }

      if (pathStr.length > 4096) {
        return false; // Linux/Unix路径长度限制
      }

      return true;
    } catch (error) {
      return false;
    }
  }

  /**
   * 规范化路径
   * @param inputPath 输入路径
   * @returns 规范化后的路径
   */
  static normalizePath(inputPath: string): string {
    if (!inputPath) return '';
    
    try {
      return path.resolve(inputPath.trim());
    } catch (error) {
      logger.error('路径规范化失败:', error);
      return inputPath.trim();
    }
  }

  /**
   * 检查路径是否可写
   * @param dirPath 目录路径
   * @returns 是否可写
   */
  static async isWritable(dirPath: string): Promise<boolean> {
    try {
      await fs.access(dirPath, fs.constants.W_OK);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * 检查路径是否可读
   * @param dirPath 目录路径
   * @returns 是否可读
   */
  static async isReadable(dirPath: string): Promise<boolean> {
    try {
      await fs.access(dirPath, fs.constants.R_OK);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * 确保目录存在（如果需要的话创建）
   * @param dirPath 目录路径
   * @returns 是否成功
   */
  static async ensureDirectory(dirPath: string): Promise<boolean> {
    try {
      await fs.ensureDir(dirPath);
      return true;
    } catch (error) {
      logger.error(`创建目录失败 ${dirPath}:`, error);
      return false;
    }
  }

  /**
   * 检查两个路径是否相同
   * @param path1 路径1
   * @param path2 路径2
   * @returns 是否相同
   */
  static isSamePath(path1: string, path2: string): boolean {
    try {
      const normalized1 = path.resolve(path1.trim());
      const normalized2 = path.resolve(path2.trim());
      return normalized1 === normalized2;
    } catch {
      return false;
    }
  }
}

File: backend/src/utils/path_validator.py
"""
路径验证工具
Path Validator Utility
"""

import os
from pathlib import Path
from models.path_cache import PathValidationResult
from core.logger import logger


class PathValidator:
    """路径验证工具类"""
    
    @staticmethod
    async def validate_path(path_str: str, must_exist: bool = True) -> PathValidationResult:
        """
        验证路径有效性
        
        Args:
            path_str: 要验证的路径字符串
            must_exist: 是否必须存在
            
        Returns:
            PathValidationResult: 验证结果
        """
        try:
            # 规范化路径
            normalized_path = PathValidator.normalize_path(path_str)
            absolute_path = Path(normalized_path).resolve()
            
            # 检查路径是否存在
            exists = absolute_path.exists()
            is_directory = absolute_path.is_dir() if exists else False
            
            # 验证逻辑
            if must_exist and not exists:
                return PathValidationResult(
                    is_valid=False,
                    error=f"路径不存在: {absolute_path}",
                    exists=False,
                    is_directory=False,
                    absolute_path=str(absolute_path)
                )
            
            if exists and not is_directory:
                return PathValidationResult(
                    is_valid=False,
                    error=f"路径不是目录: {absolute_path}",
                    exists=True,
                    is_directory=False,
                    absolute_path=str(absolute_path)
                )
            
            return PathValidationResult(
                is_valid=True,
                error=None,
                exists=exists,
                is_directory=is_directory,
                absolute_path=str(absolute_path)
            )
            
        except Exception as e:
            logger.error(f"路径验证失败: {e}")
            return PathValidationResult(
                is_valid=False,
                error=f"路径验证失败: {str(e)}",
                exists=False,
                is_directory=False,
                absolute_path=None
            )
    
    @staticmethod
    def normalize_path(path_str: str) -> str:
        """
        规范化路径
        
        Args:
            path_str: 原始路径字符串
            
        Returns:
            str: 规范化后的路径
        """
        if not path_str:
            return ""
        
        # 去除首尾空白
        path_str = path_str.strip()
        
        # 替换反斜杠为正斜杠（Windows兼容）
        path_str = path_str.replace('\\', '/')
        
        # 规范化路径
        normalized = os.path.normpath(path_str)
        
        return normalized
    
    @staticmethod
    async def ensure_directory(path_str: str) -> bool:
        """
        确保目录存在，如果不存在则创建
        
        Args:
            path_str: 目录路径
            
        Returns:
            bool: 是否成功创建或已存在
        """
        try:
            path_obj = Path(path_str)
            path_obj.mkdir(parents=True, exist_ok=True)
            return True
        except Exception as e:
            logger.error(f"创建目录失败 {path_str}: {e}")
            return False
